### PR TITLE
Fix some components not accounting for final hold notes in osu!mania beatmaps

### LINKED ISSUE #21354: Mania SV changes are not applied in the last long note of a map
### Type

Game behaviour

### Bug description

Scroll speed changes may not be applied to the last long note(s) in a mania map in very specific circumstances: If a note that has shorter length than the last note (last as in, the note that gets released last) is placed after or at the same time as the last note, scroll speed changes are not applied for the remainder of the last note.

I suspect this to be caused by code that doesn't take into account that the last object's endtime may not necessarily be the last point in the map. Most likely this does not just apply to mania maps, but having multiple notes overlapping is unusual in other rulesets.

Example map that shows this behavior in its last LN: https://osu.ppy.sh/beatmapsets/706478#mania/1506035
This map can be "fixed" by removing the two non-LN notes at 5:06:863

### Screenshots or videos

How it currently looks like in lazer (note the constant speed of the last LN):

https://user-images.githubusercontent.com/35152647/202979887-5342802d-6a2d-4bde-a4a2-bbbf864d86d7.mp4

And how it should look like (and how it looks in stable): 

https://user-images.githubusercontent.com/35152647/202980066-c3758a6d-64ef-4f1f-a18a-5e5224a0625b.mp4

### Version

2022.1117.0

### Logs

-

### PR DESCRIPTION
**Pull request by peppy**
Closes #21354.

### CODE REVIEW

**Comment by ikkue** [comment]
I think there's a typo in the PR title

comopnents → components

File: osu.Game/Beatmaps/Beatmap.cs [hunk 1]
```diff
@@ -81,9 +81,12 @@ public Beatmap()
 
         public double GetMostCommonBeatLength()
         {
+            if (!HitObjects.Any())
+                return ControlPointInfo.TimingPoints.LastOrDefault()?.Time ?? 0;
```
Comment (bdach) [inline osu.Game/Beatmaps/Beatmap.cs:85]: This is not a correct value to return from this method. The returned value should be a beat length. This is the absolute time of the last timing point.

The value cannot be early-returned, it must be assigned to `lastTime` if the beatmap has no objects.

File: osu.Game/Rulesets/UI/Scrolling/DrawableScrollingRuleset.cs [hunk 2]
```diff
@@ -114,7 +114,7 @@ private void load()
                     break;
             }
 
-            double lastObjectTime = Objects.LastOrDefault()?.GetEndTime() ?? double.MaxValue;
+            double lastObjectTime = Beatmap.HitObjects.Any() ? Beatmap.GetLastObjectTime() : 0;
```
Comment (bdach) [inline osu.Game/Rulesets/UI/Scrolling/DrawableScrollingRuleset.cs:117]: Behaviour has changed. Prior to this change, the fallback value was `double.MaxValue`, but is now 0.

```suggestion
            double lastObjectTime = Beatmap.HitObjects.Any() ? Beatmap.GetLastObjectTime() : double.MaxValue;
```

File: osu.Game/Screens/Edit/Editor.cs [hunk 3]
```diff
@@ -538,12 +537,10 @@ protected override bool OnKeyDown(KeyDownEvent e)
...
                     // Note that in osu-stable subsequent presses when at track end won't return to last object.
                     // This has intentionally been changed to make it more useful.
-                    double? lastObjectTime = editorBeatmap.HitObjects.LastOrDefault()?.GetEndTime();
-
-                    if (lastObjectTime == null || clock.CurrentTime == lastObjectTime)
+                    if (!editorBeatmap.HitObjects.Any() || clock.CurrentTime == editorBeatmap.GetLastObjectTime())
                         clock.Seek(clock.TrackLength);
                     else
-                        clock.Seek(lastObjectTime.Value);
+                        clock.Seek(editorBeatmap.GetLastObjectTime());
```
Comment (bdach) [inline osu.Game/Screens/Edit/Editor.cs:543]: If the condition in the `if` evaluates to `false`, then `GetLastObjectTime()` will be computed twice. Which seems wasteful given that the method is annotated as expensive.

**Review by bdach — CHANGES_REQUESTED** [review]

**Comment by bdach** [comment]
More places that would benefit from this sort of fix:

https://github.com/ppy/osu/blob/c5bc071c4be4b0528960106a6de3944128915367/osu.Game/Rulesets/Mods/ModTimeRamp.cs#L74
https://github.com/ppy/osu/blob/master/osu.Game/Screens/Play/HUD/SongProgress.cs#L40-L41

File: osu.Game/Beatmaps/Beatmap.cs (hunk 1 above)
Comment (peppy) [reply osu.Game/Beatmaps/Beatmap.cs:85]: Yeah, apologies, I made this change in a hurry and didn't get to self-reviewing.

File: osu.Game/Screens/Edit/Editor.cs (hunk 3 above)
Comment (peppy) [reply osu.Game/Screens/Edit/Editor.cs:543]: I did notice this, but i didn't see it as a huge concern. Will move out to a local.

**Comment by peppy** [comment]
Can't be applied to the `SongProgress` usage because we don't have a `Beatmap` reference there.

**Review by smoogipoo — APPROVED** [review]
I wonder on the API of retrieving the last time rather than just the last object, but I suppose we can cross that bridge when we get there.

**Pull request merged**